### PR TITLE
fix retention/immutable API issue

### LINKED ISSUE #18118: Cannot create a retention policy for project
bad request: project 7 already has retention policy

### PR DESCRIPTION
**Pull request by zyyw**
Signed-off-by: Shengwen Yu <yshengwen@vmware.com>

Thank you for contributing to Harbor!

# Comprehensive Summary of your change

# Issue being fixed
Fixes #18118 

Please indicate you've done the following:
- [ ] Well Written Title and Summary of the PR
- [ ] Label the PR as needed. "release-note/ignore-for-release, release-note/new-feature, release-note/update, release-note/enhancement, release-note/community, release-note/breaking-change, release-note/docs, release-note/infra, release-note/deprecation"
- [ ] Accepted the DCO. Commits without the DCO will delay acceptance.
- [ ] Made sure tests are passing and test coverage is added if needed.
- [ ] Considered the docs impact and opened a new docs issue or PR with docs changes if needed in [website repository](https://github.com/goharbor/website).

### CODE REVIEW

File: src/server/v2.0/handler/project.go [hunk 1]
```diff
@@ -677,6 +692,11 @@ func (a *projectAPI) getProject(ctx context.Context, projectNameOrID interface{}
 }
 
 func (a *projectAPI) validateProjectReq(ctx context.Context, req *models.ProjectReq) error {
+	// validate RetentionID
```
Comment (wy65701436) [inline src/server/v2.0/handler/project.go:695]: remove the nonsense comment

File: src/server/v2.0/handler/project.go [hunk 2]
```diff
@@ -557,6 +560,18 @@ func (a *projectAPI) UpdateProject(ctx context.Context, params operation.UpdateP
 		log.Warningf("failed to call JSONCopy on project metadata when UpdateProject, error: %v", err)
 	}
 
+	// validate retention_id
+	if value, ok := p.Metadata["retention_id"]; ok {
```
Comment (wy65701436) [inline src/server/v2.0/handler/project.go:564]: update `value` to an meaningful name.

**Review by chlins — APPROVED** [review]
lgtm

File: src/server/v2.0/handler/project.go [hunk 3]
```diff
@@ -677,6 +692,10 @@ func (a *projectAPI) getProject(ctx context.Context, projectNameOrID interface{}
 }
 
 func (a *projectAPI) validateProjectReq(ctx context.Context, req *models.ProjectReq) error {
+	if req.Metadata.RetentionID != nil && *req.Metadata.RetentionID != "" {
```
Comment (stonezdj) [inline src/server/v2.0/handler/project.go:695]: Replice with len(req.Metadata.RetentionID)>0

File: src/server/v2.0/handler/project.go [hunk 4]
```diff
@@ -557,6 +560,18 @@ func (a *projectAPI) UpdateProject(ctx context.Context, params operation.UpdateP
 		log.Warningf("failed to call JSONCopy on project metadata when UpdateProject, error: %v", err)
 	}
 
+	// validate retention_id
+	if ridParam, ok := p.Metadata["retention_id"]; ok {
+		md, err := a.metadataMgr.Get(ctx, p.ProjectID)
+		if err != nil {
+			return a.SendError(ctx, err)
+		}
+		if rid, ok := md["retention_id"]; !ok || rid != ridParam {
```
Comment (stonezdj) [inline src/server/v2.0/handler/project.go:569]: Replace 
```
if rid, ok := md["retention_id"]; !ok || rid != ridParam
```
With 
```
if ridParam != md["retention_id"]
```
will make it simpler.

Comment (zyyw) [reply src/server/v2.0/handler/project.go:569]: it cannot be replace that way. if ridParam is passed in as an empty string and this project doesn't have any retention policy,  we should send a 400 code, while `if ridParam != md["retention_id"]` cannot achieve that.

File: src/server/v2.0/handler/project.go (hunk 3 above)
Comment (zyyw) [reply src/server/v2.0/handler/project.go:695]: These two are logically the same, while the original code 
```
if req.Metadata.RetentionID != nil && *req.Metadata.RetentionID != ""
```
tells the logic more straight forward as the code is read, "the passing in retention_id is not given or given an empty string". But the 
```
len(req.Metadata.RetentionID)>0
```
have an indirect logic that len(nil) = 0.

**Review by stonezdj — APPROVED** [review]

File: src/server/v2.0/handler/project.go [hunk 5]
```diff
@@ -209,6 +209,9 @@ func (a *projectAPI) CreateProject(ctx context.Context, params operation.CreateP
 		log.Warningf("failed to call JSONCopy on project metadata when CreateProject, error: %v", err)
 	}
 
+	// remove retention_id from p
+	delete(p.Metadata, "retention_id")
```
Comment (chlins) [inline src/server/v2.0/handler/project.go:213]: How about move this logic to project controller? leaves the handler to do the authorization/validation(from API framework level) and  the let controller do the business logic(erase the retention ID).

**Comment by Razniak** [comment]
Hey, the problem persists, I'm still able to delete the policy via API, and then I can't add a new one because the old one is still assigned to the project. I am using Harbor version 2.8.2

**Comment by Razniak** [comment]
In addition, by creating a project via the API, I can assign 1 id retention policy to many projects, which I did, now the policy doesn't work and I can't remove it because it can't be added later, I haven't tried the workarround you mention yet, but it seems to me it's a bug.

**Comment by OrlinVasilev** [comment]
@Razniak thanks for your comment! @zyyw can you help out please :)

**Pull request merged**